### DISCUSSION TITLE
ID typescript type

### DISCUSSION THREAD

**Discussion by jamiehodge** [hayes/pothos#105] [Q&A] [answered]
It appears as though `ID` maps to `string | number` by default. By adding the following to the builder types, I appear to be able to constrain this to `string`. The question is whether it is safe at runtime or whether we risk being passed a `number`.

```graphql
ID: {
  Input: string;
  Output: string;
};
```

**Comment by hayes** [comment, marked as answer]
I generally re-define it as string like you mentioned, but the default is the way it is because the graphql server implementations (and graphql js itself) allow numbers to be passed through for ids.  I like ids always being strings, but its not something that is enforced by default, so if you want it to be type-safe you need to find a way convert your inputs.  I've done this a couple different ways, but I don't have a method I feel comfortable recommending.  It can be done with a custom girapqhl plugin that wraps resolvers (the way globalId input parsing works in the relay plugin is the closest example I can think of).  It can be handled by a custom plugin for your server (apollo and graphql helix both have ways to add custom validation logic), or I believe you can overwrite a method on the ID scalar so that it either throws or coerces numbers to strings.

Each of these methods have downsides.

* A giraphql plugin is complex and introduces a small amount of runtime overhead.
* server plugins are also a bit complicated, and I am less familiar with the details of how this would work.
* modifying the ID scalar is easy, but feels hacky, and you need to be sure it is the same copy of the scalar used in you schema (no duplicate graphql modules in your node_modules directory).  using something like yarn resolutions can make this a little more reliable.

Sorry I don't have a simple answer, but I hope that helps.

**Reply by GauBen** [reply]
> modifying the ID scalar is easy

I'm not sure about the *easy* part, but I managed to do it like this:

```ts

// Parse GraphQL IDs as numbers
const id = (
  builder.configStore.getInputTypeRef("ID") as BuiltinScalarRef<number, string>
).type;

id.parseValue = (value: unknown) => {
  const coerced = Number(value);
  if (isNaN(coerced) || !isFinite(coerced))
    throw new GraphQLError("Expected ID to be a numeric.");
  return coerced;
};

id.parseLiteral = (node) => {
  if (node.kind !== Kind.INT && node.kind !== Kind.STRING)
    throw new GraphQLError("Expected ID to be a numeric.");
  return id.parseValue(node.value);
};
```

You may also use this to force IDs to be valid UUIDs or CUIDs.
